### ISSUE TITLE
alicloud_ecs_disk.instance_id is required by API

### ISSUE THREAD

**Issue by aschleifer** [aliyun/terraform-provider-alicloud#3593]
Hi,

I'm using version 1.123.1 of the provider on terraform 0.13.7 and I have the following issue.

I'm trying to create a data disk which later can be assigned to a kubernetes deployment as persistent volume.

For this we use `alicloud_ecs_disk` to create a disk. `terraform validate` and `terraform plan` both look fine, but when we run `terraform apply` we get this error:

```
Error: [ERROR] terraform-provider-alicloud/alicloud/resource_alicloud_ecs_disk.go:471: Resource d-gw88qg7r54lqrc1vbz5k ModifyDiskChargeType Failed!!! [SDK alibaba-cloud-sdk-go ERROR]:
SDKError:
   Code: InvalidParameter
   Message: code: 400, The specified parameter "InstanceId" is not valid. request id: 34D9ECB7-68AA-4A54-9D33-D46902744319
   Data: {"Code":"InvalidParameter","HostId":"ecs.eu-central-1.aliyuncs.com","Message":"The specified parameter \"InstanceId\" is not valid.","Recommend":"https://error-center.aliyun.com/status/search?Keyword=InvalidParameter\u0026source=PopGw","RequestId":"34D9ECB7-68AA-4A54-9D33-D46902744319"}
  on main.tf line 9, in resource "alicloud_ecs_disk" "this":
   9: resource "alicloud_ecs_disk" "this" {
```

The problem is we don't have an ECS instance to attach the disk to, since it will be used in the cluster later. also according to the documentation https://registry.terraform.io/providers/aliyun/alicloud/latest/docs/resources/ecs_disk#instance_id we are not allowed to set `instance_id` and `zone_id`.

Both of these are listed as "Optional" but when we don't set the `zone_id` we also get an error that it has to be set.

**Comment by xiaozhu36** [comment]
HI @aschleifer According to the OpenAPI ModifyDiskChargeType requirement, the instance_id must be set when you want to update the instance charge type.
The "Optional" means there is no need to set when you want to creating a new disk.

**Comment by aschleifer** [comment]
> HI @aschleifer According to the OpenAPI ModifyDiskChargeType requirement, the instance_id must be set when you want to update the instance charge type.
> The "Optional" means there is no need to set when you want to creating a new disk.

Hello,

I don't have an instance to work with this disk. I want to use the disk as data disk for persistent volumes in kuberntes.

Do you mean this also happens if I change the `payment_type` of the `alicloud_ecs_disk` resource? But how does this make sense? From my understanding that `payment_type` is related to the disk, not any ECS instance.

**Comment by arafato** [comment]
@xiaozhu36 We have identified the following bug:

`alicloud_ecs_disk` allows the creation of a disk when only `zone_id` is set. This is fine and is also in line with the API documentation which requires either zoneId or instanceId.

However, the call to `ModifyDiskChargeType`(which seems to be always executed by `alicloud_ecs_disk`) requires `instance_id` to be set always! This leads to an error message by Terraform (but still creates the resource). 
 
This behaviour needs to be fixed.

**Comment by aschleifer** [comment]
Also the `alicloud_ecs_disk` resource is not atomic. We found out by accident today that the disk actually was created but the `terraform applied` had failed with the above error.

**Comment by xiaozhu36** [comment]
HI @arafato @aschleifer The API `ModifyDiskChargeType` requires an instance_id. I will push the product team and let you know if there is some progress.

**Comment by xiaozhu36** [comment]
HI @aschleifer Can you share the configuration about this resource? If the configuration contains the `payment_type`, this issue will happen.

**Comment by aschleifer** [comment]
```
resource "alicloud_ecs_disk" "this" {
  count = var.disk_count

  zone_id              = var.zone_id
  category             = var.category
  delete_auto_snapshot = var.delete_auto_snapshot
  delete_with_instance = var.delete_with_instance
  description          = var.description
  disk_name            = var.disk_name
  enable_auto_snapshot = true
  payment_type         = "PayAsYouGo"
  size                 = var.size
  type                 = "online"
}
```

As you can see I'm setting `payment_type` to "PayAsYouGo", even so it is the default. In my book this shouldn't affect anything

**Comment by aschleifer** [comment]
@xiaozhu36 you are right. If I define the `payment_type` (even to "PayAsYouGo") then I run into these issues. 

This is highly annoying and should be fixed. Only if I actually change the PaymentType I expect calls to happen to modify it. 

https://github.com/aliyun/terraform-provider-alicloud/blob/master/alicloud/resource_alicloud_ecs_disk.go#L447-L450 it looks like this is the issue as this block will be triggered for a new disk if i set `payment_type` to any value, no matter if it is different from the default or not. This sets the `update` flag to `true` which will trigger a call to "ModifyDiskChargeType" (which needs an `instance_id`).

**Comment by xiaozhu36** [comment]
@aschleifer Yes, I know, and I will fix it in the next version.

**Comment by xiaozhu36** [comment]
HI @aschleifer This issue has been fixes by version 1.124.1. Please have a check.
